background/services/provider-bridge/index.ts

```diff
@@ -51,36 +58,108 @@ export default class ProviderBridgeService extends BaseService<Events> {
     super()
 
     browser.runtime.onConnect.addListener(async (port) => {
-      if (port.name === EXTERNAL_PORT_NAME) {
-        port.onMessage.addListener(async (event) => {
-          logger.log(
-            `background: request payload: ${JSON.stringify(event.request)}`
-          )
-          const response = {
-            id: event.id,
-            result: await this.routeContentScriptRPCRequest(
-              event.request.method,
-              event.request.params
-            ),
-          }
-          logger.log("background response:", response)
-
-          port.postMessage(response)
-        })
+      if (port.name === EXTERNAL_PORT_NAME && port.sender?.url) {
+        const listener = this.onMessageListener(
+          port as Required<browser.Runtime.Port>
+        )
+        port.onMessage.addListener(listener)
+        // TODO: store port with listener to handle cleanup
       }
     })
 
     // TODO: on internal provider handlers connect, disconnect, account change, network change
   }
 
-  routeContentScriptRPCRequest(
+  onMessageListener(
+    port: Required<browser.Runtime.Port>
+  ): (event: PortRequestEvent) => Promise<void> {
+    const url = port.sender.url as string
+    const favIconUrl = port.sender.tab?.favIconUrl ?? ""
+
+    return async (event: PortRequestEvent) => {
+      // a port: browser.Runtime.Port is passed into this function as a 2nd argument by the port.onMessage.addEventListener.
+      // This contradicts the MDN documentation so better not to rely on it.
+      logger.log(
+        `background: request payload: ${JSON.stringify(event.request)}`
+      )
+
+      const response: PortResponseEvent = { id: event.id, result: [] }
+
+      if (
+        event.request.method === "eth_requestAccounts" &&
+        !(await this.checkPermission(url))
+      ) {
+        const permissionRequest: PermissionRequest = {
+          url,
+          favIconUrl,
+          state: "request",
+        }
+
+        const blockUntilUserAction = await this.requestPermission(
+          permissionRequest
+        )
+        await blockUntilUserAction
+
+        if (await this.checkPermission(url)) {
```

Comment:
If you make the permissions request resolve with the actual permission outcome, you can tighten this up:

```suggestion
        if (await this.requestPermission(
          permissionRequest
        )) {
```

Reply:
Loved the explicitness and the `checkPermission` is desinged to work from indexeddb and `await this.checkPermission(url)` had the additional meta meaning,  that the info has been persisted into the indexeddb and has been retrieved properly so there was no place for race conditions. 

But I also like your solution, let's give it a go and return back to this if needed.

Reply:
Not sure I follow, but will hold my tongue until the IndexedDB bits land.